src/core/public/context/context.ts

```diff
@@ -0,0 +1,165 @@
+   * @param plugin - The plugin this provider is associated with. If `undefined`, provider gets access to all provided
+   *                 context keys.
+   * @returns The `ContextContainer` for method chaining.
+   */
+  register<TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, TProviderParameters>,
+    plugin?: PluginName
+  ): this;
+
+  /**
+   * Create a new context.
+   *
+   * @param plugin - The plugin the context will be exposed to.
+   * @param baseContext - The initial context for the given handler.
+   * @param contextArgs - Additional parameters to call providers with.
+   * @returns A Promise for the new context object.
+   */
+  createContext(
+    plugin: PluginName,
+    baseContext: Partial<TContext>,
+    ...contextArgs: TProviderParameters
+  ): Promise<TContext>;
+}
+
+/** @internal */
+export class ContextContainerImpl<TContext extends {}, TProviderParameters extends any[] = []>
+  implements ContextContainer<TContext, TProviderParameters> {
+  /** Used to map contexts to their providers and associated plugin  */
+  private readonly contextProviders = new Map<
+    keyof TContext,
+    {
+      provider: ContextProvider<TContext, keyof TContext, TProviderParameters>;
+      plugin?: PluginName;
+    }
+  >();
+  /** Used to keep track of which plugins registered which contexts for dependency resolution. */
+  private readonly contextNamesByPlugin = new Map<PluginName, Array<keyof TContext>>();
+
+  /**
+   * @param pluginDependencies - A map of plugins to an array of their dependencies.
+   */
+  constructor(private readonly pluginDependencies: ReadonlyMap<PluginName, PluginName[]>) {}
+
+  public register<TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, TProviderParameters>,
+    plugin?: PluginName
+  ): this {
+    if (this.contextProviders.has(contextName)) {
+      throw new Error(`Context provider for ${contextName} has already been registered.`);
+    } else {
+      this.contextProviders.set(contextName, { provider, plugin });
+
+      if (plugin) {
+        this.contextNamesByPlugin.set(plugin, [
+          ...(this.contextNamesByPlugin.get(plugin) || []),
+          contextName,
+        ]);
+      }
+
+      return this;
+    }
+  }
+
+  public async createContext(
+    pluginName: PluginName,
+    baseContext: Partial<TContext> = {},
+    ...contextArgs: TProviderParameters
+  ): Promise<TContext> {
+    const contextNamesForPlugin = (plug: PluginName): Array<keyof TContext> => {
+      const pluginDeps = this.pluginDependencies.get(plug)!;
+      return flatten(pluginDeps.map(p => this.contextNamesByPlugin.get(p)!));
+    };
+
+    const neededContexts = new Set([
+      ...(this.contextNamesByPlugin.get(pluginName) || []),
+      ...contextNamesForPlugin(pluginName),
+    ]);
+
+    return [...this.contextProviders.entries()]
```

Comment:
Spreading a Map will automatically splat this into entries:

```ts
return [...this.contextProviders]
  .filter([contextName, wrappedProvider]) =>
```

Reply:
Nice, didn't know that 😄